core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.data;
+
+import static java.math.BigInteger.ZERO;
+
+import static io.parsingdata.metal.Util.checkNotNegative;
+import static io.parsingdata.metal.Util.checkNotNull;
+import static io.parsingdata.metal.data.Slice.createFromSource;
+
+import java.math.BigInteger;
+import java.util.Objects;
+import java.util.Optional;
+
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.value.ValueExpression;
+import io.parsingdata.metal.token.Token;
+
+public class ParseState {
```

Comment:
Perhaps make final?

Reply:
It's probably worth considering in general which classes should be made `final`: #223.